driver-core/src/main/com/mongodb/client/model/expressions/EntryExpression.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+
+public interface EntryExpression<T extends Expression> extends Expression {
```

Comment:
Can you clarify the pros and cons of introducing this type vs just using DocumentExpression?  I'm concerned that it's not flexible enough for all use cases, e.g. what if I want to treat each entry as a document (say, to add another field to each one).

Reply:
If we have a separate `Entry` type, we can:

1. better understand what entrySet and asMap are doing
2. provide better discoverability for iteration/mapping over keys or values
3. use Entries to get/set the keys and values in a terse and safe way (because we cannot type in the wrong field name).
4. be assured that whatever is an Entry can be used to build a map without making a mistake.

If we need to add a field, we can map Entries into Documents (and vice-versa), which would make it more clear what we are doing.

Reply:
Maybe a better question is why we need `DocumentExpression`.  What's the different between that and `MapExpression<Expression>`.  Forgive me if I'm being dense, but I want to make sure we _really_ need to introduce another concept into the API.  (`DocumentExpression` is now reminding me of a Java `Map` with no generic type).

Reply:
Ping on this question.

Reply:
Since we are supporting types elsewhere, it seems important to support `T' on `Map<T>`. DocumentExpression's `getBoolean` should not be merged into `Map`, because `MapExpression<String> strMap` would allow `strMap.getBoolean`. So we should have two types (map<T> and "record").

I think the open question was whether we should make Record extend Map<Ex>. The methods do seem to work out (for example, `unset` will not break records because everything is immutable, and it will convert a record into a map<ex>). However, I have trouble making sense of it - for example, I don't understand what it would mean for Java (or even JS) objects to extend Map. It seems convenient, but incorrect. It seems better to allow users to convert records to map<ex> (though not vice-versa). But in that case, it seems users should have used a type of map<ex> to begin with.

On the other hand, having two non-inheriting types would complicate aggregation stages, unless perhaps they shared a super-type.

Reply:
The last part of my [comment below](https://github.com/mongodb/mongo-java-driver/pull/1054#discussion_r1048972876) might address some aspect of the issue we are discussing here.

Reply:
`isMapOr` seems appropriate for some use cases but awkward for when I already have a document expression.  How would one set the value of the "foo" field on the "current" document to `1`?  Something like `current().isMapOr(ofEmptyMap()).set("foo", of(1))` seems awkward

Reply:
If the types are non-inheriting, this would be:
```
current().set("foo", of(1))
```
There is no need to convert current to a map first.